### ISSUE TITLE
Save/Open (including Save/Load Session) file dialogs don't accept forward slashes in 'File name:' box

### ISSUE THREAD

**Issue by sasumner** [notepad-plus-plus/notepad-plus-plus#9374]
### Description of the Issue
Post integration of PR #9286, I noticed that the SaveAs dialog used when saving a session does not support forward slashing in the pathname.  It would be nice if such slashing were allowed.

### Steps to Reproduce the Issue
1. _File_ > _Save Session...
2. Attempt to specify a (valid) full path with forward slashes in its name into the _SaveAs_ dialog's _File name_ box:
![image](https://user-images.githubusercontent.com/30118311/104074022-25e02180-51dd-11eb-9a6a-80d1212730c9.png)
3. Press the _Save_ button and observe the error:
![image](https://user-images.githubusercontent.com/30118311/104074077-3ee8d280-51dd-11eb-89d9-271ae15b01b8.png)
4. Change the slashes to backward ones:
![image](https://user-images.githubusercontent.com/30118311/104074123-4f994880-51dd-11eb-92fd-d01d6a9eb9e3.png)
5. Observe session file saves without error into the destination pathname

### Expected Behavior
To be able to use either forward ( `/` ) or backward ( `\` ) slashes in pathnames in the SaveAs (and Open) dialogs.

### Actual Behavior
Unable to use forward slashes.

### Debug Information
Notepad++ source after this commit: https://github.com/notepad-plus-plus/notepad-plus-plus/commit/b9f498f0412bc1806ee238627b6099d26e76e9cf

**Comment by mere-human** [comment]
I would like to work on this issue. Could it be assigned to me?

**Comment by mere-human** [comment]
It seems that common item dialogs don't allow a custom validation of the file name.
There are several features that I've tried so far:
- [SetOptions](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifiledialog-setoptions): FOS_NOVALIDATE, FOS_NOTESTFILECREATE
- [IFileDialogEvents ](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ifiledialogevents) to receive notifications of events within the dialog (a warning is shown before OnFileOk).

Neither of those works. The warning is shown nevertheless:
![image](https://user-images.githubusercontent.com/9664141/104655915-15d9ad80-56c7-11eb-8193-6a329fefdee9.png)

It seems the advanced features like file name validation is hidden intentionally by design.

So now I'm trying more tricky ways. 
When we need custom file names like the ones with a forward slash, then we'll do workarounds. 
My idea:
1. hide the standard "OK" button
![image](https://user-images.githubusercontent.com/9664141/104655775-dca13d80-56c6-11eb-88de-1115395b30af.png)
2. add a new custom "OK" button
3. do validation when the custom "OK" button is pressed
4. after all is done, send the "pressed" message to the standard button and handle the results.

**Comment by sasumner** [comment]
> So now I'm trying more tricky ways.
> When we need custom file names like the ones with a forward slash, then we'll do workarounds.

I suppose waiting on this experimentation (to see if winXP still needs to be supported) is the smart thing at the moment.

**Comment by mere-human** [comment]
@sasumner 
I am still experimenting.
So far I came up with an extra button "Navigate".
![image](https://user-images.githubusercontent.com/9664141/104786169-e3968180-5794-11eb-92f5-fbf727ac86d9.png)
It transforms a Unix-like path into a Windows path. If the name is a file path, we open the file. If the name is a directory, we go into it.
This is how it works:

https://user-images.githubusercontent.com/9664141/104785988-80a4ea80-5794-11eb-972c-39bbd7b628bb.mp4

Workaround with hiding the "Save" button is even trickier than I thought.
At first, I thought I'd add a custom button via [IFileDialogCustomize::AddPushButton](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifiledialogcustomize-addpushbutton) and then change its position to be the same as "Save" button's. 
But turns out, the custom button is contained within a "visual group":
![image](https://user-images.githubusercontent.com/9664141/104786353-64ee1400-5795-11eb-97ca-36730457f228.png)

It restricts the button's position so it is being clipped when moved.
![image](https://user-images.githubusercontent.com/9664141/104786288-2fe1c180-5795-11eb-8b61-203296ef9e42.png)

Now I am thinking about adding a custom child button, not through the IFileDialogCustomize.

How important is it to have that Unix-like path feature?

**Comment by mere-human** [comment]
Just for the reference: 
As far as I can see, the original Unix-like path feature request was mainly for the File Open dialog.
Issues: #2438, #3840
Fixed in PR #3948
Community forum topics:
https://community.notepad-plus-plus.org/topic/10573/feature-request-support-unix-slashes-in-file-open-dialog
https://community.notepad-plus-plus.org/topic/15942/opening-unix-style-file-paths
https://community.notepad-plus-plus.org/topic/16127/support-of-unix-like-paths-and-file-names-in-file-open-dialog-possible

**Comment by sasumner** [comment]
>How important is it to have that Unix-like path feature?

To me?  Not important at all.  :-)

However, I stated before (I think) that it would be nice to be able to get rid of the _Use new style..._ preference setting, and the only reason I can see that it still exists relates to being able to use `/` in paths.

**Comment by sasumner** [comment]
>"Navigate"...It transforms a Unix-like path into a Windows path. If the name is a file path, we open the file. If the name is a directory, we go into it.

The term "Navigate" isn't that intuitive for converting paths.

I'm liking the brainstorming you're doing, however; keep it up! :-)

**Comment by donho** [comment]
@mere-human 
>> "Navigate"...It transforms a Unix-like path into a Windows path. If the name is a file path, we open the file. If the name is a directory, we go into it.
> The term "Navigate" isn't that intuitive for converting paths.

+1 for @sasumner.
From my viewpoint, "Navigate" button shouldn't exist. Users should be able to provide whichever slash in the `File name` field like: `c:/myFolder\mySubFolder/myFile.txt`, then it will translate to the windows format: `c:\myFolder\mySubFolder\myFile.txt`.

>> How important is it to have that Unix-like path feature?
> To me? Not important at all. :-)

It's important because it's existed since longtime ago and this feature is used by people.
If we remove this feature, then there will be a lot of complains on the community forum and opened issues on Github.
So the statement of @sasumner is not true :)


> However, I stated before (I think) that it would be nice to be able to get rid of the Use new style... preference setting, and the only reason I can see that it still exists relates to being able to use / in paths.

I could not agree more.

**Comment by mere-human** [comment]
Ok, thanks for the input.
I've come up with a solution that doesn't require any new ugly buttons.

**Comment by donho** [comment]
Sorry I just modified the issue title because it's more interesting to have all the Save/Open dialog to have such capacity.

**Comment by mere-human** [comment]
> Sorry I just modified the issue title because it's more interesting to have all the Save/Open dialog to have such capacity.
> (I was not aware of the title)

Maybe it makes sense to create a separate issue to extend the scope of CustomFileDialog to Open/Save dialogs.
Because this way we are trying 2 things at once:
1. Add support of a forward slash path to CustomFileDIalog (this issue)
2. Use CustomFileDialog for open/save

I don't mind doing it in the scope of a single issue though.

**Comment by donho** [comment]
I see your point.
But if we keep the commit separately for the reviewing, I don't see the problem.
The thing is to make sure this feature can be implemented globally, so we can avoid the situation that only some/few dialogs can have this capacity - in this case, we have to revert the PR I've merged because the behaviour is not homogeneous.
